### ISSUE TITLE
NATS can not work with istio

### ISSUE THREAD

**Issue by YoninL** [istio/old_issues_repo#338]
**Is this a BUG or FEATURE REQUEST?:**
Not sure, maybe a BUG.

**Did you review https://istio.io/help/ and existing issues to identify if this is already solved or being worked on?**:
Yes.

**Bug:**
Y

**What Version of Istio and Kubernetes are you using, where did you get Istio from, Installation details**
```
istioctl version
0.7.1
kubectl version
1.9.5
```

**Is Istio Auth enabled or not ?**
Did you install the stable istio.yaml, istio-auth.yaml....  or if using the Helm chart please provide full command line input.
No.

**What happened**:
After [NATS](https://nats.io) cluster being istio injected, it doesn't behavior correctly, also services in istio mesh can't access NATS server outside of the mesh (but still in the same k8s cluster).

**What you expected to happen:**
NATS can be istiofied, or istio can support NATS, services inside istio mesh can access NATS cluster outside of mesh in someway.
**How to reproduce it:**
1. inside k8s, build a NATS cluster, ref https://github.com/pires/kubernetes-nats-cluster
2. use NATS go client library inside a istiofied pod in the same k8s cluster with NATS cluster to connect to the cluster, you'll find it said "no servers available for connection" (istio is injected with --includeIPRanges to cover whole private network)
3. if you use NATS go client library in a non-istiofied pod, you can connect to the cluster
4. if you move the NATS cluster out of k8s or out of the IP range included by istio, no matter if your services pod in istio-injected or not, the services can always connect to NATS cluster

[nats-istioproxy.log](https://github.com/istio/issues/files/2004963/nats-istioproxy.log)
[nats-server.log](https://github.com/istio/issues/files/2004964/nats-server.log)

[nats-client-test-connection.go](https://github.com/istio/issues/files/2004979/nats-client-test-connection.go.txt)

**Comment by lisaxxfxu** [comment]
Hi istio team, do you have any comment for this issue?

**Comment by thangchung** [comment]
Currently, I am starting to have a look at how NATS and Istio can work together. And I also want to build a POC about this, but maybe we have got the issue. Istio team, do you have any action on this?

**Comment by eastpav** [comment]
Also, istio can not work with EMQ!

**Comment by louiscryan** [comment]
Im not a NATS expert so some clarification here would be helpful in diagnosis

Does NATS use UDP or other non-TCP protocol ? I think not but want to be sure.

Is NATS dynamically allocating ports. If so how many ? Again it seems not but want to be sure.

Does NATS use its own name resolution mechanism for looking up IPs. E.g. something that would cause it to bypass the kubernetes VIPs ? This could interfere with Istio's ability to figure out which service the client is trying to talk to.

It could also be related to some timing issues pre 0.8 where Envoy doesn't receive its network configuration as fast as some clients need.

Could you retest with 0.8 to see if this is resolved?

**Comment by YoninL** [comment]
Thanks I'll take time to test it. NATS is using TCP protocol, but its url start with nats://, not sure if this is the cause.

**Comment by costinm** [comment]
I looked at NATS in the past ( I still track GCP/webpush like systems ) -
it has a text-based protocol and a protobuf-over-wire
streaming protocol. I didn't see anything in the protocol definition that
wouldn't work with Istio for single server - except the expectation
of long-lived connections, which may require adjusting istio configs.
https://nats.io/documentation/internals/nats-protocol-demo/ has a nice
example that can help debugging.

Clustering is more interesting - it seemed to rely on (stable) IP
addresses, and I haven't seen any k8s config to see how they
handle that - stateful sets or using a VIP or something else. It's not
dealing with same use cases as webpush, just pubsub,
servers just need to find a seed IP and exchange subscriptions, not
subscribers.

**Comment by YoninL** [comment]
istio 0.8.0 + nats 1.2.0
Tried with code based on https://github.com/pires/kubernetes-nats-cluster, client can't connect to it in two cases:
1. if nats cluster is not istio-injected, error is "authorization violation"
2. if nats cluster is istio-injected, error is "i/o timeout"

But I tried with https://github.com/nats-io/nats-operator, the cluster is up and can be connected, as the operator still in alpha status, I will put cluster outside of k8s until it's stable.
I didn't inject istio into the pods created by nats-operator as currently it's working.
